File: backend/internal/domain/roles.go
package domain

import "gorm.io/gorm"

// Role represents a user role in the system.
type Role struct {
	gorm.Model
	Name        string `gorm:"uniqueIndex;not null"`
	Description string
	IsSystem    bool         `gorm:"default:false"` // System roles (Admin, Manager) cannot be deleted
	Permissions []Permission `gorm:"many2many:role_permissions;"`
}

// Permission represents a granular system permission.
type Permission struct {
	gorm.Model
	Name        string `gorm:"uniqueIndex;not null"` // e.g., "inventory.read"
	Description string
	Group       string // e.g., "Inventory", "CRM"
}

// RolePermission is the join table for Roles and Permissions.
// Explicit definition allows for custom fields if needed later.
// Note: GORM handles many2many automatically, but define this if we need hooks or extra data.
// For now, we rely on GORM's automatic join table or explicit if we want to query it directly easier.
// Explicit model is safer for seeding and direct manipulation.
type RolePermission struct {
	RoleID       uint `gorm:"primaryKey"`
	PermissionID uint `gorm:"primaryKey"`
}

File: client/src/lib/api/roles.ts
import api from '$lib/api';

export interface Permission {
    ID: number;
    Name: string;
    Description: string;
    Group: string;
}

export interface Role {
    ID: number;
    CreatedAt: string;
    UpdatedAt: string;
    DeletedAt: string | null;
    Name: string;
    Description: string;
    IsSystem: boolean;
    Permissions: Permission[];
}

export const rolesApi = {
    list: async () => {
        const response = await api.get('/roles');
        return response.data as Role[];
    },

    create: async (data: { name: string; description: string }) => {
        const response = await api.post('/roles', data);
        return response.data as Role;
    },

    update: async (id: number, data: { name: string; description: string }) => {
        const response = await api.put(`/roles/${id}`, data);
        return response.data as Role;
    },

    delete: async (id: number) => {
        await api.delete(`/roles/${id}`);
    },

    listPermissions: async () => {
        const response = await api.get('/permissions');
        return response.data as Permission[];
    },

    updatePermissions: async (roleId: number, permissionIds: number[]) => {
        await api.put(
            `/roles/${roleId}/permissions`,
            { permission_ids: permissionIds }
        );
    }
};
